### DISCUSSION TITLE
Assembly of signed hash into PDF

### DISCUSSION THREAD

**Discussion by ruimelodev** [MatthiasValvekens/pyHanko#90] [Q&A]
I was given the task of assembling a digital signature on a PDF document. It's a simple task, but the more I research, the more confused I get about how to do it. Sorry if this is very basic stuff.

Essentially, I have a PDF document that I converted to base 64 and sent to an API of an Online Certifier. This API responds with a string of a signed hash of the document in SHA256.
Now, I need to create the assembly process and embed this signed hash in the PDF document.

I'm just trying to sign the document with this hash, I don't intend to add an image with a visual signature to the document.

I've checked all the signature tests here, but it seemed to me that they all make use of data that I don't have, such as .cert, .pem, passwords, etc... I reiterate that the only thing I'm given is a string of the signed hash of my PDF document.
I'd be grateful if someone could help me understand where is the functionality to accomplish this inside pyHanko.

Thank you very much for your time

**Comment by MatthiasValvekens** [comment]
Hi @ruimelodev,


The answer to your question depends on exactly what the server is returning. Do you just get back a base64 blob? Or does the service also return separate data (in particular, the signer's certificate)? How large is the payload you get back? Is it on the order of 0 to ~500-ish bytes, or a few kilobytes?


If all you have is a single blob, then I would bet that it is a CMS object that has been DER-encoded and passed through base64. These can be used in pyHanko by way of the `PdfCMSEmbedder` API, but there's a catch: to sign a PDF document, you don't actually hash the full document, but you first "prepare" it for signing by allocating a container for the signature, and then compute a hash of everything _outside_ that region. This is the same regardless of whether you're using visible or invisible signatures. If you're simply passing the entire document to be signed to the service that does the signing, then whatever comes out won't be usable in a PDF signature.

There are ways around that problem (even if the signing service isn't geared towards PDF signatures), but I'm afraid there's not much I can say without more info about the kind of service that you're using to produce the signatures. At the very least, you'd need to know whether the service returns "raw" signatures or full CMS objects. If you can't get access to any signing certificate, then chances are it's the latter, but still. Do you have any specs on hand for the service that you're using?

**Reply by 2017jw** [reply]
I have a similar case.
I have got a usb stick with cert and private key both in it. We can not access the private key, but we can use the usb stick and the program the vendor provided to log in and sign something with a short password. The vendor provides com/java api to sign given data without our acess to the private key. So can I still use the PdfCMSEmbedder you mentioned to get my pdf signed?

**Reply by MatthiasValvekens** [reply]
Hi @2017jw,

That's a good question, and it's certainly related in some way, but can you please create a separate discussion thread for your question? :)

(I'm asking because unlike most remote signing services, hardware tokens like yours _usually_ come with a PKCS#11 interface, for which pyHanko already has native integrations. Let's hash that out in another thread.)

**Reply by ruimelodev** [reply]
Thank you @MatthiasValvekens and sorry for the time I took to get back to this.

Answering your questions, I receive a base64 blob and the payload is between 0 and 500 bytes.

With your help, I discovered that I'm wrongfully hashing the full document, so all the hashes I've received are wrong.
![image](https://user-images.githubusercontent.com/16306125/166856799-d04ab015-643b-4fbe-93b6-e76efbf9241e.png)
I've found this illustration pointing to the areas that should be hashed inside the PDF, but I'm clueless about how to extract this specific parts, does pyHanko offer any functionality to do this?

Regarding your question about what the service returns, I'm contacting the service to ask more about this, I'll share here as soon as I have that info. 
Thank you

**Reply by ruimelodev** [reply]
I've received a response and apparently the returned hash is a result of the document hash signed with the Certificate.

**Reply by MatthiasValvekens** [reply]
>  I've found this illustration pointing to the areas that should be hashed inside the PDF, but I'm clueless about how to extract this specific parts, does pyHanko offer any functionality to do this?

Well, that's kind of the whole point of using a PDF signing tool like this ;).

> I've received a response and apparently the returned hash is a result of the document hash signed with the Certificate.

OK, reading your original question in light of this, I think there's some confusion about basic assumptions. Were you sending entire PDF documents, or hashes thereof, base64-encoded to the signing service? Because unless the service in question also _returns_ entire PDF documents, a workflow of that type is impossible to implement reliably. As such, it's safe to assume that that's not what's happening. In addition, you mentioned not having access to any certificates, which implies that you can't be expected to put together the CMS container yourself.

Conclusion: from your description, I think the signing service you're using expects you to pass in a document hash, and you get back a complete CMS container (DER-encoded and then wrapped in base64, probably). This is perhaps not the most common mode of operation, but it's by no means rare. If you don't have access to the signer's (public) certificate, then your best bet is [the low-level `PdfCMSEmbedder` API](https://pyhanko.readthedocs.io/en/latest/lib-guide/signing.html#the-low-level-pdfcmsembedder-api). I won't list off all the caveats here, but please carefully review the docs for `PdfCMSEmbedder`, because it's very easy to use wrongly.

You'll notice that the sample code in the linked section ends with this (comments stripped for brevity):

```python
prep_digest, output = cms_writer.send(
    cms_embedder.SigIOSetup(md_algorithm=md_algorithm, in_place=True)
)

signer: signers.SimpleSigner = FROM_CA
cms_bytes = signer.sign(
    data_digest=prep_digest.document_digest,
    digest_algorithm=md_algorithm, timestamp=timestamp
).dump()
sig_contents = cms_writer.send(cms_bytes)

# The (signed) output document is in `output` now.
```

You'll want to replace that with something like this:

```python
prep_digest, output = cms_writer.send(
    cms_embedder.SigIOSetup(md_algorithm=md_algorithm, in_place=True)
)

# Note: feel free to serialise/deserialise prep_digest and the contents of the output
# handle while you wait for a response from the service.
cms_bytes = b64decode(call_signing_service(b64encode(prep_digest.document_digest)))
prep_digest.fill_reserved_region(output, cms_bytes)
```

Hope that helps. For specific information regarding exactly what information the signing service supplies and returns, please get in touch with your vendor's support staff. I can't help you with that.

**Reply by ruimelodev** [reply]
EDIT: updated the print screen with the relevant errors and refined my questions.

Thank you @MatthiasValvekens 

With your help, I've created my own version of the `call_signing_service` function and I was able to generate a signed document.
However, when I open the file in Adobe Reader, the signature is considered invalid:
<img width="502" alt="image" src="https://user-images.githubusercontent.com/16306125/168611723-882ea254-bf77-4e53-a542-fc2e3e9278f9.png">

My questions are:
- For my tests, I'm using a blank PDF file that I've just created, is it possible that the signed hash is considered invalid because I'm not sending a valid document hash when creating `prep_digest`?
- I might be wrong, but I was expecting to see "Signed by x" instead of "Certified by x", is this defined by the signed hash that I'm receiving?

Other questions:
- What happens when I try to sign a document that already has a certificate or a signatures?
- In the event of dealing with invalid documents, does pyHanko have a way do validate this before attempting the signature assembly? I found ways to validate the signature inside pyHanko, but none for document validation.

Below is the code I'm using:
```python
root_path = os.path.abspath(os.path.dirname(__file__))
input_buf = BytesIO(open(root_path + "/document_orig.pdf", "rb").read())
w = IncrementalPdfFileWriter(input_buf)

cms_writer = cms_embedder.PdfCMSEmbedder().write_cms(field_name="Signature", writer=w)
sig_field_ref = next(cms_writer)

assert sig_field_ref.get_object()["/T"] == "Signature"

timestamp = datetime.now(tz=tzlocal.get_localzone())
sig_obj = signers.SignatureObject(timestamp=timestamp, bytes_reserved=8192)

md_algorithm = "sha256"
cms_writer.send(
    cms_embedder.SigObjSetup(
        sig_placeholder=sig_obj,
        mdp_setup=cms_embedder.SigMDPSetup(
            md_algorithm=md_algorithm,
            certify=True,
            docmdp_perms=fields.MDPPerm.NO_CHANGES,
        ),
    )
)

prep_digest, output = cms_writer.send(
    cms_embedder.SigIOSetup(md_algorithm=md_algorithm, in_place=True)
)

encoded_digest = b64encode(prep_digest.document_digest)
response = call_signing_service(encoded_digest.decode('UTF-8'))

cms_bytes = b64decode(json.loads(response.text)[0]["hashSig"])
prep_digest.fill_reserved_region(output, cms_bytes)

with open(f"output-{time.time()}.pdf", "wb") as f:
    f.write(output.getbuffer())
```

I don't know if it helps, but the vendor's support staff has sent to me an example of a Java implementation using iText.
I would be very grateful, if you could take a quick look to that example, just to understand if I'm not miscommunicating  something for this assembly process.

Thank you very much for all your help.
[TOTPSignPDF.zip](https://github.com/MatthiasValvekens/pyHanko/files/8696380/TOTPSignPDF.zip)

**Reply by MatthiasValvekens** [reply]
Hi @ruimelodev,

That sample code actually answers pretty much all of the questions I had. Anyway, the gist of it is the following:

 - I assumed that because you didn't have access to any certificates, the signing service would be returning full CMS objects. Apparently it isn't: you're really just getting a signature computed over the hash that you sent to them. The service expects the client to put together the CMS.
 
 - The problem: you need certs to put together a CMS container. How do you get one of those? Well, there's a comment in TOTPRestController that says that you can fetch the certificate belonging to your TOTPID by making a call to the `getCertificateByTOTPID` endpoint.

So, while we were working off mistaken assumptions, this is good news, really. If you can prefetch the certificate, you can make use of pyHanko's higher-level API without worrying about the details of how to assemble the CMS and embed it into the PDF. To get something up and running quick, you'll want to implement a custom `Signer` class. You can get some inspiration from this example using AWS KMS: https://pyhanko.readthedocs.io/en/latest/lib-guide/adv-examples.html#async-aws-kms.

Basically, you'll implement the `async_sign_raw` method in a custom `Signer` subclass, and your code to sign a document will look something like this:

```python
w = IncrementalPdfFileWriter(input_buf)

 # this is where you call the getCertificateByTOTPID endpoint
cert = fetch_cert_from_service(...)

# instantiate your custom signer with the cert you obtained + any other parameters
my_signer = MyAwesomeCustomTOTPSigner(signing_cert=cert, ...)
meta = signers.PdfSignatureMetadata(
    field_name='MySignatureFieldName'
 )
with open('output.pdf', 'wb') as outf:
    signers.sign_pdf(w, meta, signer=my_signer, output=outf)
```

Once you get the basic stuff working, you can start thinking about making it more efficient using [the interrupted signing API](https://pyhanko.readthedocs.io/en/latest/lib-guide/signing.html#interrupted-signing).


To answer your other questions:

 - Using a blank test doc should be fine.
 - What will happen when you try to sign a document that already contains signatures depends on several factors. First, unless you sign in append mode, you'll mess up any existing signatures. That's taken care of by using `IncrementalPdfFileWriter`, so you should be OK there. The other factors are the permission settings on prior signatures in the document. You generally can't control those, but by default adding more signatures is allowed. That said, the process for enforcing these rules is _not at all_ standardised, so be prepared to occasionally run into interoperability problems.
 - The "Certified by X" thing is because you passed in `certify=True` to the `PdfSignatureMetadata` object. You'll want to drop that unless you want a certification signature :) (those are a little special)
 - No, pyHanko does not and will never perform extensive PDF file format validation beyond what it needs to work with signatures safely. The PDF standard is a massively long document (the most recent edition is ~1000 pages), and writing software to validate all its rules is a _very_ complicated task that goes way beyond what pyHanko was designed to do.

Granted, pyHanko is already relatively opinionated in that (on the whole) it rejects more invalid documents than most general-purpose PDF processors. But I wouldn't call that "validation" per se.


Hope that helps :)


EDIT approx. a year later: Actually, the `signing_cert` parameter is now also no longer required from the start in many high-level APIs, as of 0.14.0.

**Reply by fedegalan** [reply]
Dear @ruimelodev
I have the same problem, were you able to solve it? thank you

**Reply by ruimelodev** [reply]
Hi @MatthiasValvekens, 

This project had to be postponed and I'm now picking it up again.
With the help of your last message, I was able to implement a basic script to sign a document. 
   
However, I'm receiving an error when I try to load the certificate provided by the Signing Service:
```
ValueError: Field "signature_value" is missing from structure
    while parsing asn1crypto.x509.Certificate
```
  
This is the code I'm using to load the script:
```python
cert = x509.Certificate.load(b64decode(fetch_cert()["cert_64"]))

class TOTPSigner(Signer):
    def __init__(self, signing_cert: x509.Certificate):
        self.signing_cert = signing_cert

    async def async_sign_raw(self, data: bytes, digest_algorithm: str, dry_run=False) -> bytes:
        ...

my_signer = TOTPSigner(signing_cert=cert)
meta = signers.PdfSignatureMetadata(field_name='sig')
with open('output.pdf', 'wb') as outf:
    signers.sign_pdf(w, meta, signer=my_signer, output=outf)
```

This is an example of the certificate I'm receiving from the Signing Service without it being b64encoded:
[Signing Certificate.zip](https://github.com/MatthiasValvekens/pyHanko/files/10238942/Signing.Certificate.zip)

I've tried converting the certificate to .pem and then using the function `load_cert_from_pemder` but received the same error.
I believe I'm missing something but I'm a bit lost about what it is.

Thank you

**Reply by MatthiasValvekens** [reply]
Hi @ruimelodev,

That payload is...interesting. First off, it's not a DER-encoded X.509 cert, but actually a CMS `SignedData` message with a "dummy" signature computed over (signed attributes involving a hash of) the empty string. The signature itself appears to be produced by the intermediate CA, for what it's worth.

[Side note: While it used to be somewhat common to transfer certificate bundles in DER using this type of trick, common practice is to leave the `signerInfos` field completely empty in that case. It's puzzling that they included an actual signature that appears to serve no useful purpose, since the certs embedded in the payload aren't even covered by the signature, and the signed attributes are nothing special either.]


The good news is that your certificate is indeed somewhere in there. The bad news is that extracting it consistently will require some trickery, since it's stored together with two CA certs in an ASN.1 set (=> you can't assume anything about the order).

Anyway, here's an (untested) piece of code to help you get started:

```python
from asn1crypto import cms


def find_cert():
    data = b64decode(fetch())
 
    ci = cms.ContentInfo.load(data)

    for cert_choice in ci['content']['certificates']:
        cert = cert_choice.chosen
        if is_this_my_cert(cert.subject.native):
             return cert
    raise ValueError
```

This basically loops through the list until it finds a cert with a subject matching some criterion, assuming that you know what the subject of your certificate is supposed to look like (e.g. by `organization_name` or `common_name`). These appear to be test certs, so the process for the "real thing" is probably slightly different, but you get the picture.

**Comment by fedegalan** [comment]
Dear @ruimelodev 
I have the same problem, were you able to solve it? thank you

**Reply by ruimelodev** [reply]
Hi @fedegalan, I'm still working on a solution for this. I'll post it here once it's done, but I don't have an estimate at the moment.
